README.md
# non_linear_slider
A Flutter package that provides a Non Linear Slider Widget. This is a custom
Slider that works equal than the native one but instead of defining only a `min` and
a `max` value, user will be able to define as many intervals as it wants. This way slider
will have zones where its value increase/decrease quicker or slower than others.

## Example
The next example shows a Non Linear Slider where 3 intervals have been defined. 

- First interval goes from 10 to 100 taking a 25% of the slider.
- Second interval goes from 100 to 1000 and takes the next 50%.
- Last interval goes from 1000 to 10000.

![example](https://github.com/enolcasielles/non-linear-slider/blob/master/example.gif)

## Usage
Using this Widget is pretty easy. You can pass any parameter you would pass to 
`Slider` with only one exception. Instead of passing a `min` and a `max` values 
you should pass a `List` of `NLSInterval`, representing every interval you want to 
define (the package also implements this `NLSInterval` class). Each interval has a 
`min`, a `max` and a `weight`. The `weight` parameter represents the percent you 
want this interval takes on the slider. The sum of the weights should be equal to 1.

``` dart
class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  double value = 3000;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        child: Center(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              NonLinearSlider(
                intervals: [
                  NLSInterval(10, 100, 0.25),
                  NLSInterval(100, 1000, 0.5),
                  NLSInterval(1000, 10000, 0.25),
                ],
                value: value,
                onChanged: (value) {
                  this.value = value;
                  setState(() {});
                },
                label: value.round().toString(),
                divisions: 100,
              ),
              Text(value.round().toString())
            ],
          ),
        ),
      ),
    );
  }
}
```



CHANGELOG.md
## 0.0.1 - August 2021

* Initial release

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:slider_variable_interval/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/non_linear_slider.dart
import 'package:flutter/material.dart';
import 'package:non_linear_slider/models/interval.dart';

class NonLinearSlider extends StatelessWidget {
  final List<NLSInterval> sliderIntervals = [];

  final List<NLSInterval> consumerIntervals;
  final Function(double) onChanged;
  final double value;
  final int? divisions;
  final ValueChanged<double>? onChangeStart;
  final ValueChanged<double>? onChangeEnd;
  final double? initialValue;
  final String? label;
  final Color? activeColor;
  final Color? inactiveColor;
  final MouseCursor? mouseCursor;
  final SemanticFormatterCallback? semanticFormatterCallback;
  final FocusNode? focusNode;
  final bool autofocus;

  NonLinearSlider({
    Key? key,
    required List<NLSInterval> intervals,
    required this.value,
    required this.onChanged,
    this.onChangeStart,
    this.onChangeEnd,
    this.initialValue,
    this.label,
    this.activeColor,
    this.inactiveColor,
    this.mouseCursor,
    this.semanticFormatterCallback,
    this.focusNode,
    this.autofocus = false, 
    this.divisions,
  })  : consumerIntervals = intervals,
        super(key: key) {
    _checkIntervals();
    _setIntervals();
  }

  @override
  Widget build(BuildContext context) {
    return Slider(
      value: _consumerToSlider(value),
      onChanged: (double sliderValue) {
        onChanged(_sliderToConsumer(sliderValue));
      },
      min: 0,
      max: 100,
      divisions: divisions,
      onChangeEnd: onChangeEnd,
      onChangeStart: onChangeStart,
      label: label,
      activeColor: activeColor,
      inactiveColor: inactiveColor,
      mouseCursor: mouseCursor,
      semanticFormatterCallback: semanticFormatterCallback,
      focusNode: focusNode,
      autofocus: autofocus,
    );
  }

  void _setIntervals() {
    int i = 0;
    for (final NLSInterval interval in consumerIntervals) {
      double min = i == 0 ? 0 : sliderIntervals.last.max;
      double max = min + interval.weight * 100;
      sliderIntervals.add(NLSInterval(min, max, interval.weight));
      i++;
    }
  }

  void _checkIntervals() {
    double totalValue = 0;
    for(int i=0 ; i<consumerIntervals.length ; i++) {
      final NLSInterval interval = consumerIntervals[i];
      totalValue += interval.weight;
      assert(interval.min <= interval.max, 'El mínimo de un intervalo tiene que ser menor que el máximo');
      if(i == 0) 
        continue;
      if(consumerIntervals.length > 1) 
        assert(interval.min == consumerIntervals[i-1].max, 'El minimo de un intervalo tiene que ser igual la maximo del anterior');
    }
    assert(totalValue.round() == 1, 'La suma de los porcentajes que ocupa cada intervalo ha de ser igual a 1');
  }

  double _sliderToConsumer(double sliderValue) {
    double val = consumerIntervals.first.min;
    int i = 0;
    for (final NLSInterval interval in sliderIntervals) {
      NLSInterval consumerInterval = consumerIntervals[i];
      if (interval.max < sliderValue)
        val = consumerInterval.max;
      else {
        val += (sliderValue - interval.min) *
            ((consumerInterval.max - consumerInterval.min) /
                (interval.max - interval.min));
        break;
      }
      i++;
    }
    return val;
  }

  double _consumerToSlider(double consumerValue) {
    late NLSInterval ci, si;
    int index = 0;
    for (final NLSInterval interval in consumerIntervals) {
      if (interval.max < consumerValue) {
        index++;
        continue;
      }
      ci = interval;
      si = sliderIntervals[index];
      break;
    }
    return si.min +
        (consumerValue - ci.min) * ((si.max - si.min) / (ci.max - ci.min));
  }
}


lib/variable_interval_slider.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

class SliderInterval {
  final double min;
  final double max;
  final double value;

  SliderInterval(this.min, this.max, this.value,);
}

class VariableIntervalSlider extends StatefulWidget {

  final List<SliderInterval> intervals;
  final int divisions;
  final Function(double value) onChanged;
  final ValueChanged<double>? onChangeStart;
  final ValueChanged<double>? onChangeEnd;
  final String? label;
  final Color? activeColor;
  final Color? inactiveColor;
  final MouseCursor? mouseCursor;
  final SemanticFormatterCallback? semanticFormatterCallback;
  final FocusNode? focusNode;
  final bool autofocus;

  VariableIntervalSlider({
    Key? key, 
    required this.intervals, 
    required this.divisions, 
    required this.onChanged,
    this.onChangeStart,
    this.onChangeEnd,
    this.label,
    this.activeColor,
    this.inactiveColor,
    this.mouseCursor,
    this.semanticFormatterCallback,
    this.focusNode,
    this.autofocus = false,
  }) : super(key: key) {
    _checkIntervals();
  }

  @override
  _VariableIntervalSliderState createState() => _VariableIntervalSliderState();

  void _checkIntervals() {
    for(int i=0 ; i<intervals.length ; i++) {
      final SliderInterval interval = intervals[i];
      assert(interval.min <= interval.max, 'El mínimo de un intervalo tiene que ser menor que el máximo');
      if(i == 0) 
        continue;
      if(intervals.length > 1) 
        assert(interval.min == intervals[i-1].max, 'El minimo de un intervalo tiene que ser igual la maximo del anterior');
    }
  }

}

class _VariableIntervalSliderState extends State<VariableIntervalSlider> {

  List<SliderInterval> sliderIntervals = [];

  double consumerValue = 0;
  double sliderValue = 0;

  _VariableIntervalSliderState();

  @override
  void initState() {
    _setSliderIntervals();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Slider(
      value: sliderValue, 
      min: sliderIntervals.first.min,
      max: sliderIntervals.last.max,
      onChanged: (value) {
        sliderValue = value;
        consumerValue = sliderToConsumer(sliderValue);
        print('Slider: $sliderValue'); 
        print('Consumer: $consumerValue');
        widget.onChanged(consumerValue);
        setState(() {});
      },
      divisions: widget.divisions,
      onChangeEnd: widget.onChangeEnd,
      onChangeStart: widget.onChangeStart,
      label: widget.label,
      activeColor: widget.activeColor,
      inactiveColor: widget.inactiveColor,
      mouseCursor: widget.mouseCursor,
      semanticFormatterCallback: widget.semanticFormatterCallback,
      focusNode: widget.focusNode,
      autofocus: widget.autofocus,
    );
  }

  void _setSliderIntervals() {
    sliderIntervals = [];
    int i = 0;
    for(final SliderInterval interval in widget.intervals) {
      double min = i == 0 ? widget.intervals.first.min : sliderIntervals.last.max;
      double max = min + (interval.max - interval.min) / interval.value;
      sliderIntervals.add(SliderInterval(min, max, 1));
      i++;
    }
    consumerValue = widget.intervals.first.min;
    sliderValue = sliderIntervals.first.min;
  }

  double sliderToConsumer(sliderValue) {
    double val = 0;
    int i = 0;
    for(final SliderInterval interval in sliderIntervals) {
      if(interval.max < sliderValue) 
        val += (interval.max - interval.min) * widget.intervals[i].value;
      else {
        val += (sliderValue - interval.min) * widget.intervals[i].value;
        break;
      }
      i++;
    }
    return val;
  }

}

lib/main.dart
import 'package:flutter/material.dart';
import 'package:slider_variable_interval/models/interval.dart' as nls;
import 'package:slider_variable_interval/widgets/non_linear_slider.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  double value = 3000;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        child: Center(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              NonLinearSlider(
                intervals: [
                  nls.Interval(10, 100, 0.25),
                  nls.Interval(100, 1000, 0.5),
                  nls.Interval(1000, 10000, 0.25),
                ],
                value: value,
                onChanged: (value) {
                  this.value = value;
                  setState(() {});
                },
                label: value.round().toString(),
                divisions: 100,
              ),
              Text(value.round().toString())
            ],
          ),
        ),
      ),
    );
  }
}
